File: typescript/src/prototype/prototype.test.ts
import { DataFrame } from "./prototype"

test("Clone DataFrame", () => {
  const prototype = new DataFrame("Test", {
    dataMap: new Map([
      ["name", "test"],
    ]),
  })

  const clone = prototype.clone()
  expect(clone).not.toEqual(prototype)
  expect(clone.data).not.toEqual(prototype.data)
  expect(clone.data.dataMap.get("name")).toEqual(prototype.data.dataMap.get("name"))

  clone.data.dataMap.set("version", 2)
  expect(prototype.data.dataMap.has("version")).toBeFalsy()
})

File: go/prototype/prototype_test.go
package prototype

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestDataFrame_Clone(t *testing.T) {
	// It's neither data nor frame I know, but it is not the point here, right?
	prototype := NewDataFrame("Test DataFrame",
		NewData(map[string]any{
			"name":     "Test",
			"category": "Creational",
			"pattern":  "Prototype",
		}))

	clone := prototype.Clone()
	// assert.Equal runs reflect.DeepEqual that recursively check values
	assert.Equal(t, prototype.Data.Map, clone.Data.Map)

	clone.Data.Map = map[string]any{
		"title":   "Prototype, Creational Design Pattern",
		"version": 2,
	}

	assert.NotEqual(t, prototype, clone)
	assert.NotSame(t, prototype, clone)
	assert.NotSame(t, prototype.Data, clone.Data)
	assert.NotSame(t, prototype.Data.Map, clone.Data.Map)

	assert.Equal(t, "Prototype", prototype.Data.Map["pattern"])
	assert.Equal(t, "Prototype, Creational Design Pattern", clone.Data.Map["title"])

	assert.Nil(t, clone.Data.Map["name"])
	assert.Nil(t, prototype.Data.Map["version"])

	assert.Equal(t, 2, clone.Data.Map["version"])
}
